bridge_adaptivity/module/utils.py

```diff
@@ -1,16 +1,28 @@
-from django.http import Http404
-from django.shortcuts import get_object_or_404
+from logging import getLogger
 
 from module.models import Activity
 
+log = getLogger(__name__)
+
 
 def choose_activity(sequence_item=None, sequence=None):
     sequence = sequence or sequence_item.sequence
 
     try:
-        activity_source_launch_url = sequence.group.engine.engine_driver.select_activity(sequence)
-        return get_object_or_404(Activity, collection=sequence.collection, source_launch_url=activity_source_launch_url)
-    except (IndexError, Http404):
-        sequence.completed = True
-        sequence.save()
-        return None
+        engine_choose = sequence.group.engine.engine_driver.select_activity(sequence)
+        activity_source_launch_url = engine_choose.get('source_launch_url')
+        if engine_choose.get('complete'):
+            sequence.completed = True
+            sequence.save()
+            return
+        elif activity_source_launch_url:
+            return Activity.objects.filter(
+                collection=sequence.collection, source_launch_url=activity_source_launch_url
+            ).first()
+    except AttributeError:
```

Comment:
In which place this exception can happen?

Reply:
It may happen if engine recommendation returns None object.